File: src/my_robot_controller/scripts/turtle_controller.py
#!/usr/bin/env python3

import rospy
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
from turtlesim.srv import SetPen


SERVICE_NAME = "/turtle1/set_pen"

previous_x = 0


def call_set_pen_service(r, g, b, width, off):
    try:
        set_pen = rospy.ServiceProxy(SERVICE_NAME, SetPen)
        set_pen(r, g, b, width, off)
    except rospy.ServiceException as e:
        rospy.logwarn(e)


def pose_callback(pose: Pose, publisher: rospy.Publisher):
    command = Twist()
    command.linear.x = 5.0
    command.angular.z = 0.0

    if pose.x > 9.0 or pose.x < 2.0 or pose.y > 9.0 or pose.y < 2.0:
        command.linear.x = 1.0
        command.angular.z = 1.4

    global previous_x
    if pose.x >= 5.5 and previous_x < 5.5:
        rospy.loginfo("Set color to red!")
        call_set_pen_service(255, 0, 0, 3, 0)
    elif pose.x < 5.5 and previous_x >= 5.5:
        rospy.loginfo("Set color to green!")
        call_set_pen_service(0, 255, 0, 3, 0)
    previous_x = pose.x

    publisher.publish(command)


def main():
    rospy.init_node("turtle_controller")
    rospy.wait_for_service(SERVICE_NAME)
    call_set_pen_service(255, 0, 0, 3, 0)

    publisher = rospy.Publisher("/turtle1/cmd_vel", Twist, queue_size=10)
    rospy.Subscriber("/turtle1/pose", Pose, callback=pose_callback, callback_args=publisher)

    rospy.loginfo("Node has been started")

    rospy.spin()


if __name__ == "__main__":
    main()

File: src/rosrust_controller/src/turtle_controller.rs
use std::sync::Arc;
use std::sync::Mutex;
use std::time::Duration;

use rosrust::Publisher;
use rosrust_msg::geometry_msgs;
use rosrust_msg::turtlesim;

const NODE_NAME: &str = "rosrust_turtle_controller";
const SERVICE_NAME: &str = "/turtle1/set_pen";

fn call_set_pen_service(r: u8, g: u8, b: u8, width: u8, off: u8) {
    let client = rosrust::client::<turtlesim::SetPen>(SERVICE_NAME).unwrap();
    let command = rosrust_msg::turtlesim::SetPenReq {
        r,
        g,
        b,
        width,
        off,
    };
    client.req(&command).unwrap().unwrap();
}

fn pose_callback(
    pose: &turtlesim::Pose,
    publisher: &Publisher<geometry_msgs::Twist>,
    previous_x: &Arc<Mutex<f32>>,
) {
    let mut command = geometry_msgs::Twist::default();
    command.linear.x = 5.0;
    command.angular.z = 0.0;

    if pose.x < 2.0 || pose.x > 9.0 || pose.y > 9.0 || pose.y < 2.0 {
        command.linear.x = 1.0;
        command.angular.z = 1.4;
    }

    let previous_x_val = *previous_x.lock().unwrap();

    if pose.x >= 5.5 && previous_x_val < 5.5 {
        call_set_pen_service(255, 0, 0, 3, 0);
    } else if pose.x < 5.5 && previous_x_val >= 5.5 {
        call_set_pen_service(0, 255, 0, 3, 0);
    }

    *previous_x.lock().unwrap() = pose.x;

    publisher.send(command).unwrap();
}

fn main() {
    env_logger::init();
    rosrust::init(NODE_NAME);
    rosrust::wait_for_service(SERVICE_NAME, Some(Duration::from_secs(10))).unwrap();

    let previous_x = Arc::new(Mutex::new(0.0_f32));

    let publisher = rosrust::publish::<geometry_msgs::Twist>("/turtle1/cmd_vel", 10).unwrap();
    
    // Save to a variable, even though we don't use it.
    // This is due to Rust RAII (Resource Acquisition Is Initialization) mechanism,
    // which automatically destructs the resource if it's not save to a variable.
    // By saving to a variable, the resource will not be destroyed until
    // the variable goes out of scope.
    // Similar mechanism happens with C++.
    let _subscriber = rosrust::subscribe("/turtle1/pose", 10, move |pose: turtlesim::Pose| {
        pose_callback(&pose, &publisher, &previous_x);
    })
    .unwrap();

    rosrust::ros_info!("Node has been started");

    rosrust::spin();
}
